### PR TITLE
[Time Conductor] add history and select range features

### PR DESCRIPTION
**Pull request by davetsay**
## Overview
Addresses  #2430, #3047. Supercedes #2552

- change pan axis to require alt key press and hold
- add zoom axis
  - drag to select a range within current bounds to "zoom" in to
  - zoomed range becomes new bounds
- visually highlight  on pan and on zoom
- time conductor history
  - records recently used times based on current time system
  - saves time history on bounds change
  - saves time history on time system change
  - configurable presets (ex. last hour, last day)
  - configurable number of previous history records
  - persists history for all used time systems in local storage

## Author Checklist
| | |
| --- | :---: |
| Changes address original issue? | Y |
| Unit tests included and/or updated with changes? | N/A |
| Command line build passes? | Y |
| Changes have been smoke-tested? | Y |

### CODE REVIEW

**Comment by davetsay** [comment]
@reviewers,
Known Issue: When alt is pressed the cursor changes to use the pan. You can break this behavior the way it is coded now if you hold alt, then use the mouse for non-time conductor functionality, specifically, resizing the window. I'm not sure if we need to address this behavior.

What should presets be for History?

**Comment by charlesh88** [comment]
Added some extra cheese:
- Enhanced styles for `.c-menu`;
- Added hint messaging and separator;
- Reversed displayed history array so that latest entry is first;
@davetsay Please have a look and make sure my usage of `reverse()` is kosher.

**Comment by deeptailor** [comment]
@davetsay - Can we base this out of VISTA R4.4 as well, it will have to go into R4.4.1, which does not include current master

File: src/plugins/timeConductor/ConductorHistory.vue [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    </div>
+</div>
+</template>
+
+<script>
+import toggleMixin from '../../ui/mixins/toggle-mixin';
+
+const LOCAL_STORAGE_HISTORY_KEY = 'tcHistory';
+const DEFAULT_RECORDS = 10;
+const DEFAULT_UTC_PRESETS = [
```
Comment (akhenry) [inline src/plugins/timeConductor/ConductorHistory.vue:68]: I don't think we need default presets. Useful presets are going to differ a lot by deployment.

File: src/plugins/timeConductor/ConductorHistory.vue [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+        addTimespan() {
+            const key = this.timeSystem.key;
+            let [...currentHistory] = this.history[key] || [];
+            const timespan = {
+                start: this.bounds.start,
+                end: this.bounds.end
+            };
+
+            // when choosing an existing entry, remove it and add it back as latest entry
+            // const isNotEqual = (entry) => entry.start !== this.start || entry.end !== this.end;
```
Comment (akhenry) [inline src/plugins/timeConductor/ConductorHistory.vue:166]: Let's remove this commented code.

File: src/plugins/timeConductor/ConductorHistory.vue [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+            const isNotEqual = function (entry) {
+                const start = entry.start !== this.start;
+                const end = entry.end !== this.end;
+
+                return start || end;
+            };
+            currentHistory = currentHistory.filter(isNotEqual, timespan);
```
Comment (akhenry) [inline src/plugins/timeConductor/ConductorHistory.vue:173]: Nice, I like this. If it's a duplicate, remove it and add it to the top of the stack.

File: src/plugins/timeConductor/ConductorAxis.vue [hunk 4]
```diff
@@ -134,42 +167,132 @@ export default {
...
-                    let newStart = bounds.start - percX * deltaTime;
-                    this.$emit('panAxis',{
-                        start: newStart,
-                        end: newStart + deltaTime
-                    });
-                    this.dragging = false;
-                })
-            } else {
+            if (this.dragging) {
                 console.log('Rejected drag due to RAF cap');
```
[the remark this answers is not in the document]
Comment (davetsay) [reply src/plugins/timeConductor/ConductorAxis.vue]: removed

File: src/plugins/timeConductor/ConductorAxis.vue [hunk 5]
```diff
@@ -134,42 +167,132 @@ export default {
...
             }
+
+            this.dragging = true;
+
+            requestAnimationFrame(() => {
+                this.dragX = $event.clientX;
+                this.inPanMode ? this.pan() : this.zoom();
+            });
+
+            this.dragging = false;
```
[the remark this answers is not in the document]
Comment (davetsay) [reply src/plugins/timeConductor/ConductorAxis.vue]: oops. fixed.

File: src/plugins/timeConductor/ConductorHistory.vue (hunk 2 above)
Comment (davetsay) [reply src/plugins/timeConductor/ConductorHistory.vue]: removed

**Comment by charlesh88** [comment]
Testing notes:
1. Switch the Time Conductor to Fixed mode if not already.
1. Note that a "History" menu button is displayed.
1. Make a change to the Start datetime and tab out or press Return to update the TC.
1. Click the History menu button and verify that the Start and End datetimes for the first entry in the menu match the Time Conductor's current setting.
1. Repeat steps 3 and 4, except use the End datetime.
1. Click the History menu button and make a selection from the menu other than the first item.
1. Repeat step 4.

**Comment by davetsay** [comment]
> 1. When I first ran the branch, I was in RT mode, then switched to Fixed. The presets in the History menu were not appearing:
>    Screen Shot 2020-05-11 at 4 54 15 PM
> 
> They eventually showed up later after several manipulations of the Start and End - not sure why, thought it was after panning, but that's not making them appear now. Switching to RT then back to Fix made the disappear again and I can't get them back.
> 
> 1. Seeing some very weird behavior with drag zoom in the TC:
>    Time Conductor drag zoom problem 05-11-20

This should be resolved from the latest commits to the branch.

File: src/plugins/timeConductor/ConductorHistory.vue (hunk 1 above)
Comment (davetsay) [reply src/plugins/timeConductor/ConductorHistory.vue]: removed.

**Comment by deeptailor** [comment]
@charlesh88 @akhenry - Need your approval to merge this branch

**Comment by akhenry** [comment]
Since we're in the testing week of this sprint, and this is a pretty significant chunk of new functionality, let's hold off on merging this. It's good to go though.

**Comment by deeptailor** [comment]
## Reviewer Checklist
Changes appear to address issue? Yes
Appropriate unit tests included? N/A
Code style and in-line documentation are appropriate? Yes
Commit messages meet standards? Yes
Has associated issue been labelled unverified? (only applicable if this PR closes the issue) Yes

**Pull request merged**